File: ios/Modules/Sources/Feeds/impl/HttpChannelFeedRepository.swift
import Foundation
import Net
import DB

public final class HttpChannelFeedRepository: ChannelFeedRepository {
    private let http: HttpClient
    private let dao: ChannelFeedDao

    public init(http: HttpClient, dao: ChannelFeedDao) { self.http = http; self.dao = dao }

    public func observe(_ roomId: Int64) -> AsyncStream<[ChannelFeed]> {
        AsyncStream { cont in
            let inner = Task {
                for await rows in dao.observeForRoom(roomId) { cont.yield(rows.map(\.toDomain)) }
            }
            cont.onTermination = { @Sendable _ in inner.cancel() }
        }
    }

    public func refresh(_ roomId: Int64) async -> Bool {
        do {
            let list = try await http.send(.get("api/rooms/\(roomId)/feeds"), [FeedDto].self)
            for dto in list { try await dao.upsert(dto.toRecord) }
            return true
        } catch { return false }
    }

    public func subscribe(roomId: Int64, url: String, feedType: String) async -> Bool {
        do {
            let req = try HttpRequest.postJson("api/rooms/\(roomId)/feeds",
                body: SubReq(url: url, feed_type: feedType))
            let dto = try await http.send(req, FeedDto.self)
            try await dao.upsert(dto.toRecord)
            return true
        } catch { return false }
    }

    public func unsubscribe(_ feedId: Int64) async -> Bool {
        do {
            _ = try await http.send(.delete("api/feeds/\(feedId)"))
            try await dao.delete(feedId); return true
        } catch { return false }
    }

    private struct SubReq: Encodable { let url: String; let feed_type: String }
    private struct FeedDto: Decodable {
        let id: Int64
        let room_id: Int64
        let feed_type: String
        let url: String
        let last_fetched: Int64?
        let is_active: Bool?
        var toRecord: ChannelFeedRecord {
            ChannelFeedRecord(id: id, roomId: room_id,
                              feedType: feed_type, url: url,
                              lastFetched: last_fetched,
                              isActive: is_active ?? true)
        }
    }
}

private extension ChannelFeedRecord {
    var toDomain: ChannelFeed {
        ChannelFeed(id: id, roomId: roomId, feedType: feedType, url: url,
                    lastFetched: lastFetched, isActive: isActive)
    }
}

File: android/app/src/main/java/sol/vortexx/android/feeds/impl/HttpChannelFeedRepository.kt
package sol.vortexx.android.feeds.impl

import io.ktor.client.call.body
import io.ktor.client.request.delete
import io.ktor.client.request.get
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import kotlinx.coroutines.flow.Flow
import kotlinx.serialization.Serializable
import sol.vortexx.android.db.dao.ChannelFeedDao
import sol.vortexx.android.db.entities.ChannelFeedEntity
import sol.vortexx.android.feeds.api.ChannelFeedRepository
import sol.vortexx.android.net.impl.VortexHttpClient
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class HttpChannelFeedRepository @Inject constructor(
    private val http: VortexHttpClient,
    private val dao: ChannelFeedDao,
) : ChannelFeedRepository {

    override fun observe(roomId: Long): Flow<List<ChannelFeedEntity>> = dao.observeRoom(roomId)

    override suspend fun refresh(roomId: Long): Boolean = runCatching {
        val resp = http.client.get("api/rooms/$roomId/feeds")
        if (!resp.status.isSuccess()) return@runCatching false
        resp.body<List<FeedDto>>().forEach { dto ->
            dao.upsert(ChannelFeedEntity(
                id = dto.id, roomId = dto.room_id,
                feedType = dto.feed_type, url = dto.url,
                lastFetched = dto.last_fetched, isActive = dto.is_active,
            ))
        }
        true
    }.getOrDefault(false)

    override suspend fun subscribe(roomId: Long, url: String, feedType: String): Boolean = runCatching {
        val resp = http.client.post("api/rooms/$roomId/feeds") {
            contentType(ContentType.Application.Json)
            setBody(SubReq(url = url, feed_type = feedType))
        }
        if (!resp.status.isSuccess()) return@runCatching false
        val dto = resp.body<FeedDto>()
        dao.upsert(ChannelFeedEntity(
            id = dto.id, roomId = dto.room_id,
            feedType = dto.feed_type, url = dto.url,
            lastFetched = dto.last_fetched, isActive = dto.is_active,
        ))
        true
    }.getOrDefault(false)

    override suspend fun unsubscribe(feedId: Long): Boolean = runCatching {
        val resp = http.client.delete("api/feeds/$feedId")
        if (resp.status.isSuccess()) { dao.delete(feedId); true } else false
    }.getOrDefault(false)

    @Serializable private data class SubReq(val url: String, val feed_type: String)
    @Serializable private data class FeedDto(
        val id: Long, val room_id: Long, val feed_type: String, val url: String,
        val last_fetched: Long? = null, val is_active: Boolean = true,
    )
}
